packages/plugin-react/src/index.ts

```diff
@@ -256,3 +288,7 @@ viteReact.preambleCode = preambleCode
 function interopDefault(promise: Promise<any>): Promise<any> {
   return promise.then((module) => module.default || module)
 }
+
+// overwrite for cjs require('...')() usage
+module.exports = viteReact
```

Comment:
Are you sure this is necessary? What happens without it?

Reply:
The fix is brought from https://github.com/vitejs/vite/blob/main/packages/plugin-vue/src/index.ts#L190-L192

If this line is removed, when `plugin-react` is transpiled into commonjs module, the export is:

```js
exports.default = viteReact;
```

However, in Node if we do `require('@vitejs/plugin-react')`, the return value is `{default: viteReact}` instead of `viteReact`. Node does not convert `default` import into a namespace. 

Without this hack, when user `require` the plugin instead of using `import`, he will see:

```
error when starting dev server:
TypeError: reactPlugin is not a function
    at Object.<anonymous> (/Users/pengxiao/src/vite/packages/playground/react/vite.config.ts:15:13)
    at Module._compile (node:internal/modules/cjs/loader:1108:14)
```

Reply:
Have you done any testing in ESM environment?

Might need to wrap it in `if (typeof module !== 'undefined')`

Reply:
I just tested by turning on `"type": "module"` for playground/react, it looks like transpiling with TSC is not enough to let it be used in ESM environment. The error:
```
failed to load config from /Users/stb/Documents/GitHub/vite/packages/playground/react/vite.config.ts
error when starting dev server:
/Users/stb/Documents/GitHub/vite/packages/plugin-react/dist/esm/index.js:1
import * as babel from '@babel/core';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```
It seems it is not trivial to fix it without a bundler. A related ref https://github.com/microsoft/TypeScript/issues/18442

I will try a similar approach in `plugin-vue` instead with esbuild.

Reply:
Okay, I guess it's fine we don't support ESM, since `plugin-vue` doesn't either.

Reply:
Yeah, I guess we can add esm built later if there is some real use cases.